Sources/NIOCore/AsyncChannelAdapter.swift

```diff
@@ -0,0 +1,296 @@
+        self._lock.withLockVoid {
+            switch self._state {
+            case .bufferingWithPendingRead(let buffer, _):
+                self._state = .bufferingWithoutPendingRead(buffer)
+            case .bufferingWithoutPendingRead, .waitingForBuffer:
+                preconditionFailure()
+            }
+        }
+
+        switch result {
+        case .success:
+            context.read()
+        case .failure:
+            context.close(promise: nil)
+        }
+    }
+
+    @inlinable
+    func _receiveNewEvent(_ event: InboundMessage) {
+        self._lock.withLockVoid {
+            switch self._state {
+            case .bufferingWithPendingRead(var buffer, let promise):
+                // Weird, but let's tolerate it.
+                buffer.append(event)
+                self._state = .bufferingWithPendingRead(buffer, promise)
+            case .bufferingWithoutPendingRead(var buffer):
+                buffer.append(event)
+                self._state = .bufferingWithoutPendingRead(buffer)
+            case .waitingForBuffer(let buffer, let continuation):
+                precondition(buffer.count == 0)
+                self._state = .bufferingWithoutPendingRead(buffer)
+                // DANGER: This is only safe if we aren't a custom executor. Otherwise this is
+                // v. bad. We should decide if it matters.
+                continuation.resume(returning: event)
+            }
+        }
+    }
+
+    @inlinable
+    func loadEvent() async -> InboundMessage {
+        let result: (InboundMessage, EventLoopPromise<Void>?)?
+        result = self._lock.withLock {
+            switch self._state {
+            case .bufferingWithoutPendingRead(var buffer):
+                if buffer.count > 0 {
+                    let result = buffer.removeFirst()
+                    self._state = .bufferingWithoutPendingRead(buffer)
+                    return (result, nil)
+                } else {
+                    return nil
+                }
+            case .bufferingWithPendingRead(var buffer, let promise):
+                // No matter what happens here, we need to complete this promise.
+                if buffer.count > 0 {
+                    let result = buffer.removeFirst()
+                    self._state = .bufferingWithoutPendingRead(buffer)
+                    return (result, promise)
+                } else {
+                    return nil
+                }
+            case .waitingForBuffer:
+                preconditionFailure("Should never be async nexting twice.")
+            }
+        }
+
+        if let result = result {
+            result.1?.succeed(())
+            return result.0
+        } else {
+            return await self._loadEventSlowPath()
+        }
+    }
+
+    @inlinable
+    func _loadEventSlowPath() async -> InboundMessage {
+        return await withCheckedContinuation { (continuation: CheckedContinuation<InboundMessage, Never>) in
```

Comment:
how would you recommend handling errors? Why not use `CheckedContinuation<InboundMessage, Error>` and `resume(throwing:)` in the `errorCaught`?